### PR TITLE
fix(serve): When IPv6 is not enabled, attempt to bind to IPv4 address under the same hostname.

### PR DESCRIPTION
**Pull request by Hanaasagi**
### What does this PR do?

Close: #5315. 

Please take a look at this https://github.com/oven-sh/bun/issues/5315#issuecomment-1719291283 first. About the IPv6 hosts, see the discussion in https://github.com/moby/moby/issues/35954. 

Previously, uWebSockets would iterate the results of `getaddrinfo` in order of IPv6 and IPv4 priorities, and if it failed, it would return an error immediately. This PR will attempt to bind to the IPv4 address under the same hostname when IPv6 fails (if the hostname is bound to both IPv6 and IPv4).

- [ ] Documentation or TypeScript types (it's okay to leave the rest blank in this case)
- [x] Code changes

### How did you verify your code works?

manual testing.

#### Test step 

This issue can also be reproduced on the host machine.

- index.js

```JavaScript
const net = require("net");

const server = new net.Server();

server.on("listening", () => {
  console.log("listen success");
  server.close();
});

server.on("error", err => {
  console.log(err);
  server.close();
});

server.listen(3000, "localhost");
```

Add following lines to your `/etc/hosts`.

```
127.0.0.1 localhost
::1     localhost ip6-localhost ip6-loopback
```

Then disable ipv6 on loopback interface.

```Bash
sudo sysctl net.ipv6.conf.lo.disable_ipv6=1
```

Run `index.js`, Bun 1.0.6 will print 

```
error: Failed to listen at localhost
 code: "EADDRNOTAVAIL"
```

Use this pr's patch will print "listen success".

### CODE REVIEW

File: packages/bun-usockets/src/bsd.c [hunk 1]
```diff
@@ -446,24 +446,49 @@ LIBUS_SOCKET_DESCRIPTOR bsd_create_listen_socket(const char *host, int port, int
...
 
-    if (listenFd == LIBUS_SOCKET_ERROR) {
-        freeaddrinfo(result);
-        return LIBUS_SOCKET_ERROR;
-    }
+    freeaddrinfo(result);
+    return LIBUS_SOCKET_ERROR;
+}
+
+inline LIBUS_SOCKET_DESCRIPTOR _bsd_create_listen_socket(
```
Comment (dylan-conway) [inline packages/bun-usockets/src/bsd.c]: this function needs to be declared before `bsd_create_listen_socket`

File: packages/bun-usockets/src/bsd.c [hunk 2]
```diff
@@ -446,24 +446,49 @@ LIBUS_SOCKET_DESCRIPTOR bsd_create_listen_socket(const char *host, int port, int
 
     LIBUS_SOCKET_DESCRIPTOR listenFd = LIBUS_SOCKET_ERROR;
     struct addrinfo *listenAddr;
-    for (struct addrinfo *a = result; a && listenFd == LIBUS_SOCKET_ERROR; a = a->ai_next) {
+    for (struct addrinfo *a = result; a != NULL; a = a->ai_next) {
         if (a->ai_family == AF_INET6) {
             listenFd = bsd_create_socket(a->ai_family, a->ai_socktype, a->ai_protocol);
             listenAddr = a;
+
+            if (_bsd_create_listen_socket(listenFd, listenAddr, port, options) != LIBUS_SOCKET_ERROR) {
```
Comment (dylan-conway) [inline packages/bun-usockets/src/bsd.c]: should we also have a check for `listenFd != LIBUS_SOCKET_ERROR`?

Comment (dylan-conway) [reply packages/bun-usockets/src/bsd.c]: before calling the function?

File: packages/bun-usockets/src/bsd.c (hunk 1 above)
Comment (Hanaasagi) [reply packages/bun-usockets/src/bsd.c]: Fixed.

File: packages/bun-usockets/src/bsd.c (hunk 2 above)
Comment (Hanaasagi) [reply packages/bun-usockets/src/bsd.c]: Yes, fixed.

File: packages/bun-usockets/src/bsd.c [hunk 3]
```diff
@@ -446,61 +488,47 @@ LIBUS_SOCKET_DESCRIPTOR bsd_create_listen_socket(const char *host, int port, int
...
     LIBUS_SOCKET_DESCRIPTOR listenFd = LIBUS_SOCKET_ERROR;
     struct addrinfo *listenAddr;
-    for (struct addrinfo *a = result; a && listenFd == LIBUS_SOCKET_ERROR; a = a->ai_next) {
+    for (struct addrinfo *a = result; a != NULL; a = a->ai_next) {
         if (a->ai_family == AF_INET6) {
             listenFd = bsd_create_socket(a->ai_family, a->ai_socktype, a->ai_protocol);
             listenAddr = a;
+
+            if (listenFd == LIBUS_SOCKET_ERROR) {
+                break;
```
Comment (dylan-conway) [inline packages/bun-usockets/src/bsd.c]: I think this break and lines 505, 517, and 525 should be `continue` so we can try more entries in `result`

**Comment by dylan-conway** [comment]
LGTM after last comment is addressed 👍

File: packages/bun-usockets/src/bsd.c (hunk 3 above)
Comment (Hanaasagi) [reply packages/bun-usockets/src/bsd.c]: done

**Comment by dylan-conway** [comment]
We are setting up ci to run tests for this pr, will merge once we get it working

**Comment by dylan-conway** [comment]
Thanks!

**Pull request merged**